File: WebVersion/Code/Cell.ts
class Cell {

    private maxValue: number = 255;

    public Owner: number; 
    public Value: number;

    public Row: number;
    public Col: number

    constructor(row: number, col: number) {
        this.Row = row;
        this.Col = col;
        this.Owner = 0;
        this.Value = 0;
    }

    public Clone(): Cell {
        var result = new Cell(this.Row, this.Col);
        result.Value = this.Value;
        result.Owner = this.Owner;
        return result;
    }

    public UpgradeValue(): void {

        //dit kan slimmer! je bent gewoon lui he?. even bitjes tellen man. 

        if (this.Value >= 128) {
            this.Value += 8;
            if (this.Value > this.maxValue) { this.Value = this.maxValue; }
            return;
        }
        if (this.Value >= 64) { this.Value += 7; return; }
        if (this.Value >= 32) { this.Value += 6; return; }
        if (this.Value >= 16) { this.Value += 5; return; }
        if (this.Value >= 8) { this.Value += 4; return; }
        if (this.Value >= 4) { this.Value += 3; return; }
        if (this.Value >= 2) { this.Value += 2; return; }
        if (this.Value >= 1) { this.Value += 1; return; }
    }


    public AddValue(value: number, playerID: number): void {
        if ((this.Owner == 0) || (this.Owner == playerID)) {
            this.Value += value;
            if (this.Value > this.maxValue) {
                this.Value = this.maxValue;
            }
            this.Owner = playerID;
        }
        else {
            this.Value -= value;
            if (this.Value == 0) {
                this.Owner = 0;
            }

            if (this.Value < 0) {
                this.Value *= -1;
                this.Owner = playerID;
            }
        }

    }
}

File: SharedMemLib/Cell.cs
namespace SharedMemLib
{
    public class Cell
    {
        private const int MaxValue = 255;

        public int Owner { get; set; }           //0 = niemand; 1=engine1 ; 2=engine2;
        public int Value { get; set; }
        public int Row { get; set; }
        public int Col { get; set; }

        public Cell(int row, int col)
        {
            Row = row;
            Col = col;
            Owner = 0 ;
            Value = 0;
        }

        public Cell Clone()
        {
            var result = new Cell(Row, Col)
            {
                Value = Value,
                Owner = Owner
            };

            return result;
        }

        public void UpgradeValue()
        {
            //dit kan slimmer! je bent gewoon lui he?. even bitjes tellen man. 

            if (Value >= 128)
            {
                Value += 8;
                if (Value>MaxValue) { Value = MaxValue; }

                return; 
            }

            if (Value >= 64) { Value += 7; return; }
            if (Value >= 32) { Value += 6; return; }
            if (Value >= 16) { Value += 5; return; }
            if (Value >= 8) { Value += 4; return; }
            if (Value >= 4) { Value += 3; return; }
            if (Value >= 2) { Value += 2; return; }
            if (Value >= 1) { Value += 1; return; }
        }

        
        public void AddValue(int value, int playerId)
        {
            if ((Owner == 0) || (Owner == playerId))
            {
                Value += value;
                if (Value > MaxValue)
                    Value = MaxValue;

                Owner = playerId;
            }
            else
            {
                Value -= value;
                if (Value == 0)
                    Owner = 0;

                if (Value < 0)
                {
                    Value *= -1;
                    Owner = playerId;
                }
            }
        }
    }
}
